Mathlib/Data/Seq/Defs.lean

```diff
@@ -746,6 +790,30 @@ def update (s : Seq α) (n : ℕ) (f : α → α) : Seq α where
 def set (s : Seq α) (n : ℕ) (a : α) : Seq α :=
   update s n fun _ ↦ a
 
+/-!
+### Predicates on sequences
+-/
+
+/-- `s.All p` means that the predicate `p` is true on each element of `s`. -/
+def All (s : Seq α) (p : α → Prop) : Prop := ∀ x ∈ s, p x
+
+/--
+`Pairwise R s` means that all the elements with earlier indexes are
+`R`-related to all the elements with later indexes.
+```
+Pairwise R [1, 2, 3] ↔ R 1 2 ∧ R 1 3 ∧ R 2 3
+```
+For example if `R = (· ≠ ·)` then it asserts `s` has no duplicates,
+and if `R = (· < ·)` then it asserts that `s` is (strictly) sorted.
+-/
+def Pairwise (R : α → α → Prop) (s : Seq α) : Prop :=
+  ∀ i j x y, i < j → s.get? i = .some x → s.get? j = .some y → R x y
+
+/-- `s₁.AtLeastAsLongAs s₂` means that `s₁` has at least as many elements as sequence `s₂`.
+In particular, they both may be infinite. -/
+def AtLeastAsLongAs (a : Seq α) (b : Seq β) : Prop :=
+  ∀ n, a.TerminatedAt n → b.TerminatedAt n
```

Comment:
Should we have an `ENat`-valued "length?` (using choice) so that this can just be expressed with `le`? Or are there reasons that make that less convenient?